features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/resources/ResourceDataStoreManager.kt

```diff
@@ -0,0 +1,159 @@
+    private val resourceHashesDeserializer: Deserializer<String, ResourceHashesEntry>
+) {
+    private val knownResources = Collections.synchronizedSet(mutableSetOf<String>())
+    private var storedLastUpdateDateNs = AtomicLong(System.nanoTime())
+    private var isInitialized = AtomicBoolean(false) // has init finished executing its async actions
+
+    init {
+        val onDeleteDatastoreCallback = object : DataStoreWriteCallback {
+            override fun onSuccess() {
+                finishedInitializingManager()
+            }
+
+            override fun onFailure() {
+                finishedInitializingManager()
+            }
+        }
+
+        fetchStoredResourceHashes(
+            onFetchSuccessful = lambda@{ storedEntry ->
+                val storedData = storedEntry?.data
+
+                if (storedData == null) {
+                    finishedInitializingManager()
+                    return@lambda
+                }
+
+                val lastUpdateDateNs = storedData.lastUpdateDateNs
+                val storedHashes = storedData.resourceHashes
+
+                storedLastUpdateDateNs.set(lastUpdateDateNs.toLong())
+
+                if (didDataStoreExpire()) {
+                    handleDataStoreExpired(
+                        storedHashes = storedHashes.toSet(),
+                        callback = onDeleteDatastoreCallback
+                    )
+                } else {
+                    knownResources.addAll(storedHashes)
+                    finishedInitializingManager()
+                }
+            },
+            onFetchFailure = {
+                // No need to do anything here.
+                // The next time we write to the datastore it will overwrite
+            }
+        )
+    }
+
+    internal fun isPreviouslySentResource(resourceHash: String): Boolean =
+        knownResources.contains(resourceHash)
+
+    internal fun cacheResourceHash(resourceHash: String) {
+        knownResources.add(resourceHash)
+        writeResourcesToStore()
+    }
+
+    internal fun isReady(): Boolean =
+        isInitialized.get()
+
+    // region internal
+
+    private fun finishedInitializingManager() {
+        isInitialized.set(true)
+    }
+
+    private fun writeResourcesToStore() {
+        val data = ResourceHashesEntry(
+            lastUpdateDateNs = storedLastUpdateDateNs,
+            resourceHashes = knownResources.toList()
+        )
+
+        featureSdkCore.getFeature(
+            SESSION_REPLAY_RESOURCES_FEATURE_NAME
+        )?.dataStore?.setValue(
+            data = data,
+            key = DATASTORE_HASHES_ENTRY_NAME,
+            callback = object : DataStoreWriteCallback {
+                override fun onSuccess() {
+                    // no need to do anything here
+                }
+
+                override fun onFailure() {
+                    // no need to do anything here
+                }
+            },
```

Comment:
let's maybe make the callback in `setValue` signature as optional or give a no-op value by default? Majority of use-cases won't require providing a callback, same for `removeValue`.